### DISCUSSION TITLE
Go API curl Image Attachments

### DISCUSSION THREAD

**Discussion by AlfarisiN** [netbox-community/netbox#8899] [Q&A] [answered]
Hi, 
currently im having a problem when i upload an image using extras/image-attachments/ 
i have api server build with golang, and i intend to use curl to all api/extras/image-attachments/
is there any example of how i can send my request?
i tried already, this is my code

func NewfileUploadRequest(c* gin.Context) {
	url := "http://ip/api/extras/image-attachments/"

	token := c.GetHeader("Authorization")
	//acc := c.GetHeader("Accept")
	//cont := c.GetHeader("Content-Type")
	contentType := c.PostForm("content_type")
	objectId := c.PostForm("object_id")
	imgHeight := c.PostForm("image_height")
	imgWidth := c.PostForm("image_width")
	imgName := c.PostForm("name")
	formFile, _ := c.MultipartForm()
	fFile := formFile.File["image"]

	var path = ""

	extraParams := map[string]interface{}{
		"content_type":contentType,
		"object_id":objectId,
		"name":imgName,
		"image_height":imgHeight,
		"image_width":imgWidth,
	}

	if err := MakeFolder("netboxTmp/"); err != nil {
		fmt.Println("Directory creation failed with error: " + err.Error(),"here?")
		os.Exit(1)
	}
	for _, file := range fFile {
		extension := filepath.Ext(file.Filename)
		newFileName := imgName + extension

		path = "netboxTmp/"+ newFileName
		fmt.Println("isi 2 : " + path)

		if err := c.SaveUploadedFile(file, "netboxTmp/"+newFileName); err != nil {
			log.Fatal(err)
		}
	}

	files, err := os.Open(path)
	if err != nil {
		log.Fatalf(err.Error()," Here 1")
	}
	fileContents, err := ioutil.ReadAll(files)
	if err != nil {
		log.Fatalf(err.Error(), " Here 2")
	}
	fi, err := files.Stat()
	if err != nil {
		log.Fatalf(err.Error(), " Here 3")
	}
	files.Close()
	body := &bytes.Buffer{}
	writer := multipart.NewWriter(body)
	part, err := writer.CreateFormFile("image", filepath.Base(fi.Name()))
	if err != nil {
		log.Fatalf(err.Error(), " Here 3.5")
	}
	io.Copy(part, files)
	writer.Close()
	part.Write(fileContents)

	for key, val := range extraParams {
		_ = writer.WriteField(key, val.(string))
	}
	err = writer.Close()
	if err != nil {
		log.Fatalf(err.Error(), " Here 4")
	}
	fmt.Println(body)
	req, _ := http.NewRequestWithContext(c, "POST", url, body)
	//req.Header.Add("Content-Type", "multipart/form-data")
	req.Header.Add("Authorization", token)
	res, _ := http.DefaultClient.Do(req)

	defer res.Body.Close()

	bodyGet, _ := ioutil.ReadAll(res.Body)
	output := fmt.Sprint(string(bodyGet))
	//JSON BUILDER (1st try)
	in := []byte(output)
	var raw map[string]interface{}
	if err := json.Unmarshal(in, &raw); err != nil {
		panic(err)
	}
	out, err := json.Marshal(raw)
	if err != nil {
		panic(err)
	}

	if res.StatusCode == 404 {
		c.JSON(http.StatusNotFound, out)
	}

	c.JSON(res.StatusCode, raw)
}

but when i hit the request i got 415 unsupported media type

can someone point out where i did wrong here? and is it possible to use application/json with image being converted to base64 instead?

**Comment by markkuleinio** [comment]
I'd first try with the "Content-Type: application/json" header since this is a POST call: https://netbox.readthedocs.io/en/stable/rest-api/overview/#creating-a-new-object

**Reply by AlfarisiN** [reply]
sadly didn't work, i try that already with the image converted to base64 but didn't recognize by netbox.
is it possible to receive base64 image as application/json?

**Reply by markkuleinio** [reply]
https://github.com/netbox-community/netbox/discussions/6845 does not show how to use base64 either

**Comment by AlfarisiN** [comment, marked as answer]
*Edit :
i fix my problem here

anyway, here is my code for somebody need, hope my code usefull 

func NewfileUploadRequest(c* gin.Context) {
	url := "http://ip/api/extras/image-attachments/"

	token := c.GetHeader("Authorization")
	//acc := c.GetHeader("Accept")
	//cont := c.GetHeader("Content-Type")
	contentType := c.PostForm("content_type")
	objectId := c.PostForm("object_id")
	imgHeight := c.PostForm("image_height")
	imgWidth := c.PostForm("image_width")
	imgName := c.PostForm("name")
	formFile, _ := c.MultipartForm()
	fFile := formFile.File["image"]

	var pathImg = ""
	var newFileName = ""

	extraParams := map[string]interface{}{
		"content_type":contentType,
		"object_id":objectId,
		"name":imgName,
		"image_height":imgHeight,
		"image_width":imgWidth,
	}

	if err := MakeFolder("netboxTmp/"); err != nil {
		fmt.Println("Directory creation failed with error: " + err.Error(),"here?")
		os.Exit(1)
	}
	for _, file := range fFile {
		extension := filepath.Ext(file.Filename)
		newFileName = imgName + extension

		pathImg = "netboxTmp/"+ newFileName
		fmt.Println("isi 2 : " + pathImg)

		if err := c.SaveUploadedFile(file, "netboxTmp/"+newFileName); err != nil {
			log.Fatal(err)
		}
	}
	fileDir, _ := os.Getwd()
	fileName := newFileName
	filePath := path.Join(fileDir+"/netboxTmp/", fileName)
	files, err := os.Open(filePath)
	if err != nil {
		log.Fatalf(err.Error()," Here 1")
	}
	
	defer files.Close()
	body := &bytes.Buffer{}
	writer := multipart.NewWriter(body)
	part, err := writer.CreateFormFile("image", filepath.Base(files.Name()))
	if err != nil {
		log.Fatalf(err.Error(), " Here 3.5")
	}
	io.Copy(part, files)
	writer.Close()

	for key, val := range extraParams {
		_ = writer.WriteField(key, val.(string))
	}
	err = writer.Close()
	if err != nil {
		log.Fatalf(err.Error(), " Here 4")
	}
	fmt.Println(body)
	req, _ := http.NewRequestWithContext(c, "POST", url, body)
	req.Header.Add("Content-Type", writer.FormDataContentType())
	req.Header.Add("Authorization", token)
	res, _ := http.DefaultClient.Do(req)

	defer res.Body.Close()

	bodyGet, _ := ioutil.ReadAll(res.Body)
	output := fmt.Sprint(string(bodyGet))
	//JSON BUILDER (1st try)
	in := []byte(output)
	var raw map[string]interface{}
	if err := json.Unmarshal(in, &raw); err != nil {
		panic(err)
	}
	out, err := json.Marshal(raw)
	if err != nil {
		panic(err)
	}

	if res.StatusCode == 404 {
		c.JSON(http.StatusNotFound, out)
	}

	c.JSON(res.StatusCode, raw)
}
